packages/database/src/core/PersistentConnection.ts

```diff
@@ -186,15 +195,55 @@ export class PersistentConnection extends ServerActions {
   }
 
   get(query: Query): Promise<string> {
-    const req: { [k: string]: unknown } = {
+    const deferred = new Deferred<string>();
+    const request = {
       p: query.path.toString(),
       q: query.queryObject()
     };
+    const outstandingGet = {
+      action: 'g',
+      request,
+      onComplete: (message: { [k: string]: unknown }) => {
+        const payload = message['d'] as string;
+        if (message['s'] === 'ok') {
+          this.onDataUpdate_(
+            request['p'],
+            payload,
+            /*isMerge*/ false,
+            /*tag*/ null
+          );
+          deferred.resolve(payload);
+        } else {
+          deferred.reject(payload);
+        }
+      }
+    };
+    this.outstandingGets_.push(outstandingGet);
+    this.outstandingGetCount_++;
+    const index = this.outstandingGets_.length - 1;
+
+    if (!this.connected_) {
+      const self = this;
+      setTimeout(function () {
+        const get = self.outstandingGets_[index];
+        if (get === undefined || outstandingGet !== get) {
+          return;
+        }
+        delete self.outstandingGets_[index];
+        self.outstandingGetCount_--;
+        if (self.outstandingGetCount_ === 0) {
+          self.outstandingGets_ = [];
+        }
+        self.log_('get ' + index + ' timed out on connection');
+        deferred.reject(new Error('Client is offline.'));
```

Comment:
Should we have logic that falls back to cache here? This would only trigger if there was an active listener that covers this location, but for some reason we cannot send the `get()` (e.g. we were able to connect to the backend at some point but are no longer able to). If you decide against this, you should probably add a comment to explain the difference.

Reply:
Is this the right abstraction layer to add caching logic to? I don't think the serverSyncTree is in scope here. I'd prefer to add a comment.

Reply:
Sounds good, please do add the comment.